docs/topics/functions.md

```diff
@@ -135,73 +256,139 @@ You can skip all the ones with default values:
 ```
 
+<!-- Rationale for named arguments interaction with varargs is here https://youtrack.jetbrains.com/issue/KT-52505#focus=Comments-27-6147916.0-0 -->
+
 > When calling Java functions on the JVM, you can't use the named argument syntax because Java bytecode does not
 > always preserve the names of function parameters.
 >
 {style="note"}
 
-### Unit-returning functions
-
-If a function does not return a useful value, its return type is `Unit`. `Unit` is a type with only one value - `Unit`.
-This value does not have to be returned explicitly:
+### Return types
 
-```kotlin
-fun printHello(name: String?): Unit {
-    if (name != null)
-        println("Hello $name")
-    else
-        println("Hi there!")
-    // `return Unit` or `return` is optional
-}
-```
+When you declare a function with a block body (by putting instructions within curly braces `{}`),
+you must always specify a return type explicitly.
+The only exception is when they return `Unit`,
+[in which case specifying the return type is optional](#unit-returning-functions).
 
-The `Unit` return type declaration is also optional. The above code is equivalent to:
-
-```kotlin
-fun printHello(name: String?) { ... }
-```
+Kotlin doesn't infer return types for functions with block bodies.
+Their control flow can be complex, which makes the return type unclear to the reader and sometimes even to the compiler.
+However, Kotlin can infer the return type for [single-expression functions](#single-expression-functions) if you don't specify it.
 
 ### Single-expression functions
 
-When the function body consists of a single expression, the curly braces can be omitted and the body specified after an `=` symbol:
+When the function body consists of a single expression, you can omit the curly braces and specify the body after an `=` symbol:
 
 ```kotlin
 fun double(x: Int): Int = x * 2
 ```
 
-Explicitly declaring the return type is [optional](#explicit-return-types) when this can be inferred by the compiler:
+Most of the time you don't have to explicitly declare [the return type](#return-types):
 
 ```kotlin
+// Compiler infers that the function returns Int
 fun double(x: Int) = x * 2
 ```
 
-### Explicit return types
+Compiler can run into problems inferring return types from single expressions.
+Functions that are recursive or mutually recursive (calling each other)
+and functions with typeless expressions like `fun empty() = null` always require a return type.
 
-Functions with block body must always specify return types explicitly, unless it's intended for them to return `Unit`,
-[in which case specifying the return type is optional](#unit-returning-functions).
+When you do use an inferred return type,
+make sure to check the actual result because the compiler may infer a type that is less useful to you.
+If in the example above you would like the `double()` function to return `Number` instead of `Int`, for example,
+you have to declare this explicitly.
```

Comment:
```suggestion
In the example above, if you want the `double()` function to return `Number` instead of `Int`, 
you have to declare this explicitly.
```